Refactor Multiline assert statements into exceptions

bvaisvil opened the pull request:
Hiya @peterjc -

This pull request addresses issue #1515

I hereby agree to dual licence this and any previous contributions under both
the _Biopython License Agreement_ **AND** the _BSD 3-Clause License_.

I have read the ``CONTRIBUTING.rst`` file and understand that AppVeyor and
TravisCI will be used to confirm the Biopython unit tests and ``flake8`` style
checks pass with these changes.

I have added my name to the alphabetical contributors listings in the files
``NEWS.rst`` and ``CONTRIB.rst`` as part of this pull request, am listed
already, or do not wish to be listed. (*This acknowledgement is optional.*)

bvaisvil commented:
So this failed a test in codonalign because I didn't update the corresponding test (oops). I've added a commit that updates that test and passes tests locally (I must remember to build before testing).

Should I close this pull request and make a new one or is there a way to request travis/appveyor re-run the tests?

bvaisvil commented:
Thanks for the feedback! I'll be pushing another commit soon. Good to know that the tests will run again.

peterjc commented on Bio/Application/__init__.py:
```diff
@@ -212,7 +212,9 @@ def __init__(self, cmd, **kwargs):
-                assert isinstance(p, _StaticArgument), p
+                if not isinstance(p, _StaticArgument):
+                    raise TypeError("Expected %s to be of type _StaticArgument"
+                                    % repr(p))
```
This is neater using ``%r`` rather than ``%s`` and an explicit call of ``repr(...)``

```python
raise TypeError("Expected %r to be of type _StaticArgument" % p)
```

bvaisvil replied:
That's a good point - completely forgot about `%r`. I noticed there are other explicit uses of `repr` that I think could be changed to use `%r` in [`Bio/Application/__init__.py`](https://github.com/biopython/biopython/blob/ad216e5302463c95971bb35a8cf49d6904cd064a/Bio/Application/__init__.py#L226). I could change those as well as part of this PR?

peterjc replied:
Since you are working on error messages in this file anyway, the same PR should be fine. Equally a fresh PR for that issue would also be fine if you'd prefer.

peterjc commented on Bio/Application/__init__.py:
```diff
@@ -383,7 +385,9 @@ def _check_value(self, value, name, check_function):
             is_good = check_function(value)  # May raise an exception
-            assert is_good in [0, 1, True, False]
+            if is_good not in [0, 1, True, False]:
+                raise ValueError("Result of check_function: %s is of an unexpected value"
```
Can you use ``%r`` for the Python object representation - that seems safer given we don't know what kind of value has been passed in.

peterjc commented on Bio/Blast/NCBIStandalone.py:
```diff
@@ -985,7 +985,8 @@ def title(self, line):
         parts = line.replace(" ", "").split("=")
-        assert len(parts) == 2, "Unrecognised format length line"
+        if len(parts) != 2:
+            raise ValueError("Unrecognised format length line: %s" % line)
```
Can you use ``%r`` for the Python object representation here? That tends to be more useful for seeing things like tabs and carriage returns explicitly.

peterjc commented:
Almost ready to merge now, just a little more polish. Thanks!

peterjc commented:
Are you happy and confident to deal with the merge conflict, or would you prefer I do it?

More tests are welcome too - that can be a separate pull request if you like, since I think this one is ready otherwise.

bvaisvil commented:
I'd be happy to fix the merge conflict, but I'm not sure where I should look for it? On this pull request it says 'This branch has no conflicts with the base branch'. In addition, I don't see a conflict when I merged upstream/master locally. Sorry, I'm still new to developing on github.

peterjc commented:
I'm assuming there is a problem because GitHub shows me a grey warning "This branch cannot be rebased due to conflicts" rather than a merge button.

If your copy of the master branch is not fully up to date, you may need to fetch the official repository before you can see the problem (i.e. ``git fetch origin`` or depending how you set this up, maybe ``git fetch upstream`` or similar).

bvaisvil commented:
Okay, I merged `upstream/master` into `multiline_assert` and I pushed that commit. I didn't see a conflict when I merged, however.

peterjc commented:
GitHub seems happy now. Strange that git was able to handle any merge conflict automatically, but GitHub thought manual intervention was needed.

peterjc commented:
And now GitHub says "This branch cannot be rebased due to conflicts" again. Grr. No idea why. Leave this with me, I'll sort it out one way or another - hopefully tomorrow.

I'll also add you to the ``CONTRIB.rst`` file pointing at your GitHub profile.

MarkusPiotrowski commented:
@bvaisvil Your remote master is ahead of the Biopython master. Although I'm not sure what is happening here, for this type of project it's best never to touch the master. Only keep your local and/or remote master in sync with the official Biopython master (e.g. from your master branch call ```git pull upstream master```). To update your developing branch with the recent Biopython master use (in your development branch) ```git rebase master```. There is no need to merge your development branch into your master.
Note: ```git pull upstream master``` will not work with _your_ current master, since it's ahead of the Biopython master. You first need to reset it.

bvaisvil commented:
I've rebased my master branch as @MarkusPiotrowski suggested and the `multiline_assert` branch. I did find the conflict and fixed it there. I've not pushed it yet as this was my first rebase. I think I've done it right but I'm not positive. Just wanted to check here before I went ahead and pushed it.

peterjc commented:
Sorry, I did say I'd sort this out but I was on holiday in the interim and then busy with some training courses, so I'd forgotten.

It sounds like you should be OK to update this branch - note because of the rebase you will find ``git push bvaisvil multiline_assert `` fails and you have to use ``git push bvaisvil multiline_assert --force`` as it will change the history of this branch (exact command will depend on how you named your remote repositories).

If it all goes horribly wrong (which is unlikely), I have a copy of your branch as it is now, so plan B is I can work from that.

peterjc commented:
That update to the branch has not done what I was expecting. I'll try to sort this out tomorrow (Friday 11 May).

peterjc commented:
Thanks for trying the rebase, if it had worked that would have saved me a little time - but in the end you didn't cause any extra work so don't feel guilty.

Based on the older version of your branch I had kept locally, I did the rebase and as part of that squashed a few of the commits for a simpler history (e.g. hide the changes to the alphabet asserts which I asked you to revert; combine the repr changes into once commit).

Assuming this passes the automatic checks, it should be ready to merge. Thank you for working on this!

bvaisvil commented:
Thanks for the help with the merge!  I'll keep in mind adding tests for the new exceptions I add.
